### PR TITLE
Moving common HadoopSecurityManager code to AbstractHadoopSecurityManager so that all the versions of HadoopSecurityManager can reuse this code.

### PR DESCRIPTION
**Pull request by jakhani**
For Containerization project, we want to introduce HadoopSecurityManager version 3. There is a lot of code which version 3 will need from version 2. So reuse all that code, I have created AbstractHadoopSecurityManager with code which can be reused.

### CODE REVIEW

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java [hunk 1]
```diff
@@ -0,0 +1,698 @@
...
+          getToken(userToProxyFQN, userToProxy);
+          return null;
+        }
+
+        private void getToken(final String userToProxyFQN, final String userToProxy)
+            throws InterruptedException, IOException, HadoopSecurityManagerException {
+          logger.info("Here is the props for " + HadoopSecurityManager.OBTAIN_NAMENODE_TOKEN +
+              ": " + props.getBoolean(HadoopSecurityManager.OBTAIN_NAMENODE_TOKEN));
+
+          registerAllCustomCredentials(userToProxy, props, cred, logger);
```
Comment (arvindpruthi) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java:390]: In Hadoop Security Manager 2.0, this code is called ifL ENABLE_OAUTH is false. Why the change?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java [hunk 2]
```diff
@@ -0,0 +1,698 @@
...
+  protected boolean shouldProxy;
+  protected boolean securityEnabled;
+  public static final String CHOWN = "chown";
+  public static final String CHMOD = "chmod";
+
+  // Unable to use slf4j logger as this reference is passed at many places for JobTypeManager
+  private static final Logger logger = Logger
+      .getLogger(AbstractHadoopSecurityManager.class);
+
+  public AbstractHadoopSecurityManager(final Props props) {
```
[the remark this answers is not in the document]
Comment (jakhani) [reply azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java:99]: Yes that was not needed. HadoopSecurityManager2 was already catching it.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java [hunk 3]
```diff
@@ -0,0 +1,698 @@
...
+   * Properties.
+   */
+  @Override
+  public UserGroupInformation getProxiedUser(final Props userProp)
+      throws HadoopSecurityManagerException {
+    final String userToProxy = verifySecureProperty(userProp, JobProperties.USER_TO_PROXY);
+    final UserGroupInformation user = getProxiedUser(userToProxy);
+    if (user == null) {
+      throw new HadoopSecurityManagerException(
+          "Proxy as any user in unsecured grid is not supported!");
```
[the remark this answers is not in the document]
Comment (jakhani) [reply azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java:149]: You are right. Fixed it.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java [hunk 4]
```diff
@@ -0,0 +1,698 @@
...
+  /**
+   * This method is used to get property from props object. It will throw an exception when property
+   * doesn't exist.
+   *
+   * @param props
+   * @param s
+   * @return
+   * @throws HadoopSecurityManagerException
+   */
+  private String verifySecureProperty(final Props props, final String s)
```
[the remark this answers is not in the document]
Comment (jakhani) [reply azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java:163]: Yes. This method is not used anywhere outside HadoopSecurityManager_H_2_0. That is why I changed it's scope.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java (hunk 1 above)
Comment (jakhani) [reply azkaban-hadoop-security-plugin/src/main/java/azkaban/security/AbstractHadoopSecurityManager.java:390]: I have kept that check inside registerAllCustomCredentials. I have created registerAllCustomCredentials to take all the customCredentials related part out in separate method.

**Review by arvindpruthi — APPROVED** [review]
